clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java

```diff
@@ -416,6 +418,10 @@ public Void call() throws Exception {
             executor.submit(work);
         }
 
+        // Wait for the workers to be blocked in their buffer allocations.
+        ready.await(15, TimeUnit.SECONDS);
+        Thread.sleep(10);
```

Comment:
As you mentioned, without sleep(), the pool.close() would be called immediately - reverting to the behavior prior to this change.
I think the following formation would achieve what you want to test without the need to choose a proper sleep period.
```
        ready.await(15, TimeUnit.SECONDS);

        assertEquals("Allocation shouldn't have happened yet, waiting on memory", numWorkers, completed.getCount());
        Thread.yield();
```
Verified via additional logging:
```
allocating
allocating
closing
allocated
allocated
```

Reply:
> As you mentioned, without sleep(), the pool.close() would be called immediately - reverting to the behavior prior to this change.

Not exactly. The addition of `ready` ensures that the threads have at least gotten as far as immediately before the call, which is the bulk of the setup.

> I think the following formation would achieve what you want to test without the need to choose a proper sleep period.
> 
> ```
>         ready.await(15, TimeUnit.SECONDS);
> 
>         assertEquals("Allocation shouldn't have happened yet, waiting on memory", numWorkers, completed.getCount());
>         Thread.yield();
> ```
> 
> Verified via additional logging:
> 
> ```
> allocating
> allocating
> closing
> allocated
> allocated
> ```

So keep in mind the sleep is in terms of milliseconds, where 10 ms shouldn't be an issue, as opposed to yield which can be more relaxed in whether to suspend the thread. However you'd want the yield to come before the assertion to provide more chance that the workers have made it inside the buffer pool and have been unable to make progress.

Reply:
@bdbyrne Couldn't we wait for the pool to have non-empty waiters instead of using Thread.sleep?

Reply:
Yes. Done.